// nexus/dist-esm/runtime/schema/schema.js
import * as NexusSchema from '@nexus/schema';
import chalk from 'chalk';
import { createNexusSchemaStateful } from '../../lib/nexus-schema-stateful';
import * as Process from '../../lib/process';
import * as Scalars from '../../lib/scalars';
import * as DevMode from '../dev-mode';
import { assertAppNotAssembled } from '../utils';
import { log } from './logger';
import { createSchemaSettingsManager } from './settings';
import { mapSettingsAndPluginsToNexusSchemaConfig } from './settings-mapper';
function createLazyState() {
    return {
        contextContributors: [],
        plugins: [],
        scalars: {},
    };
}
export function create(appState) {
    appState.schemaComponent = createLazyState();
    const statefulNexusSchema = createNexusSchemaStateful();
    const settings = createSchemaSettingsManager();
    const api = Object.assign(Object.assign({}, statefulNexusSchema.builders), { use(plugin) {
            assertAppNotAssembled(appState, 'app.schema.use', 'The Nexus Schema plugin you used will be ignored.');
            appState.schemaComponent.plugins.push(plugin);
        },
        addToContext(contextContributor) {
            appState.schemaComponent.contextContributors.push(contextContributor);
        },
        middleware(fn) {
            api.use(NexusSchema.plugin({
                // TODO: Do we need to expose the name property?
                name: 'local-middleware',
                onCreateFieldResolver(config) {
                    return fn(config);
                },
            }));
        } });
    return {
        public: api,
        private: {
            settings: settings,
            reset() {
                statefulNexusSchema.state.types = [];
                statefulNexusSchema.state.scalars = {};
                appState.schemaComponent.contextContributors = [];
                appState.schemaComponent.plugins = [];
                appState.schemaComponent.scalars = {};
            },
            beforeAssembly() {
                appState.schemaComponent.scalars = statefulNexusSchema.state.scalars;
            },
            assemble(plugins) {
                const nexusSchemaConfig = mapSettingsAndPluginsToNexusSchemaConfig(plugins, settings.data);
                nexusSchemaConfig.types.push(...statefulNexusSchema.state.types);
                nexusSchemaConfig.plugins.push(...appState.schemaComponent.plugins);
                const { schema, missingTypes } = NexusSchema.core.makeSchemaInternal(nexusSchemaConfig);
                return { schema, missingTypes };
            },
            checks() {
                assertNoMissingTypesDev(appState.assembled.schema, appState.assembled.missingTypes);
                // TODO: We should separate types added by the framework and the ones added by users
                if (statefulNexusSchema.state.types.length === 2 &&
                    statefulNexusSchema.state.types.every((t) => Scalars.builtinScalars[t.name] !== undefined)) {
                    log.warn(emptyExceptionMessage());
                }
            },
        },
    };
}
function emptyExceptionMessage() {
    return `Your GraphQL schema is empty. This is normal if you have not defined any GraphQL types yet. If you did however, check that your files are contained in the same directory specified in the \`rootDir\` property of your tsconfig.json file.`;
}
function assertNoMissingTypesDev(schema, missingTypes) {
    const missingTypesNames = Object.keys(missingTypes);
    if (missingTypesNames.length === 0) {
        return;
    }
    const schemaTypeMap = schema.getTypeMap();
    const schemaTypeNames = Object.keys(schemaTypeMap).filter((typeName) => !NexusSchema.core.isUnknownType(schemaTypeMap[typeName]));
    if (DevMode.isDevMode()) {
        missingTypesNames.map((typeName) => {
            const suggestions = NexusSchema.core.suggestionList(typeName, schemaTypeNames);
            let suggestionsString = '';
            if (suggestions.length > 0) {
                suggestionsString = ` Did you mean ${suggestions
                    .map((s) => `"${chalk.greenBright(s)}"`)
                    .join(', ')} ?`;
            }
            log.error(`Missing type "${chalk.redBright(typeName)}" in your GraphQL Schema.${suggestionsString}`);
        });
    }
    else {
        Process.fatal(`Missing types ${missingTypesNames.map((t) => `"${t}"`).join(', ')} in your GraphQL Schema.`, { missingTypesNames });
    }
}
